### PR TITLE
[Merged by Bors] - feat: computation of Levenshtein edit distance

### PR DESCRIPTION
**Pull request by kim-em**
Subsequent PRs will prove properties of this definition, and implement the `rw_search` tactic using it.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/EditDistance/Defs.lean [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+/-- A cost structure for Levenshtein edit distance. -/
+structure Cost (α β : Type _) (δ : Type _) where
+  delete : α → δ
+  insert : β → δ
+  substitute : α → β → δ
+
+/-- The default cost structure, for which all operations cost `1`. -/
+def defaultCost [DecidableEq α] : Cost α α ℕ :=
+  ⟨fun _ => 1, fun _ => 1, fun a b => if a = b then 0 else 1⟩
```
Comment (eric-wieser) [inline Mathlib/Data/List/EditDistance/Defs.lean:52]: I think this is worth the extra lines for clarity:
```suggestion
def defaultCost [DecidableEq α] : Cost α α ℕ where
  delete _ := 1
  insert _ := 1
  substitute a b := if a = b then 0 else 1
```

File: Mathlib/Data/List/EditDistance/Defs.lean [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+instance [DecidableEq α] : Inhabited (Cost α α ℕ) := ⟨defaultCost⟩
+
+@[simp] lemma defaultCost_delete [DecidableEq α] (a : α) : defaultCost.delete a = 1 := rfl
+@[simp] lemma defaultCost_insert [DecidableEq α] (a : α) : defaultCost.insert a = 1 := rfl
+@[simp] lemma defaultCost_substitute [DecidableEq α] (a b : α) :
+    defaultCost.substitute a b = if a = b then 0 else 1 := rfl
```
Comment (eric-wieser) [inline Mathlib/Data/List/EditDistance/Defs.lean:59]: I think `@[simps]` can generate these

File: Mathlib/Data/List/EditDistance/Defs.lean [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+compute the Levenshtein distances from each suffix of `xs` to `y :: ys`.
+
+(Note that we don't actually need to know `ys` itself here, so it is not an argument.)
+
+The return value is a list of length `x.length + 1`,
+and it is convenient for the recursive calls that we bundle this list
+with a proof that it is non-empty.
+-/
+def impl
+    (xs : List α) (y : β) (d : Σ' (r : List δ), 0 < r.length) : Σ' (r : List δ), 0 < r.length :=
```
Comment (eric-wieser) [inline Mathlib/Data/List/EditDistance/Defs.lean:76]: Why `Σ'` and not
```suggestion
    (xs : List α) (y : β) (d : {r : List δ // 0 < r.length}) : {r : List δ // 0 < r.length} :=
```

File: Mathlib/Data/List/EditDistance/Defs.lean [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    (init := ⟨[C.insert y + ds.getLast (List.length_pos.mp w)], by simp⟩)
+    (fun ⟨x, d₀, d₁⟩ ⟨r, w⟩ =>
+      ⟨min (C.delete x + r[0]) (min (C.insert y + d₀) (C.substitute x y + d₁)) :: r, by simp⟩)
+
+variable {C}
+variable (x : α) (xs : List α) (y : β) (d : δ) (ds : List δ) (w : 0 < (d :: ds).length)
+
+-- Note this lemma has an unspecified proof `w'` on the right-hand-side,
+-- which will become an extra goal when rewriting.
+theorem impl_cons (w') :
```
Comment (eric-wieser) [inline Mathlib/Data/List/EditDistance/Defs.lean:88]: Can you make the type of `w` explicit here? (and below)

File: Mathlib/Data/List/EditDistance/Defs.lean (hunk 2 above)
Comment (semorrison) [reply Mathlib/Data/List/EditDistance/Defs.lean:59]: Indeed!

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

File: Mathlib/Data/List/EditDistance/Defs.lean [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+          (C.substitute x y + levenshtein C xs ys)) :=
+  suffixLevenshtein_cons_cons_fst_get_zero _ _ _ _ _
+
+#guard
+  (suffixLevenshtein Levenshtein.defaultCost "kitten".toList "sitting".toList).1 =
+    [3, 3, 4, 5, 6, 6, 7]
+
+#guard levenshtein Levenshtein.defaultCost
+  "but our fish said, 'no! no!'".toList
+  "'put me down!' said the fish.".toList = 21
```
Comment (eric-wieser) [inline Mathlib/Data/List/EditDistance/Defs.lean:288]: Should these be in a separate test file? Should we have a style guide that gives an opinion on this?

(no need to un-bors the PR, we can clean this type of thing up another time)

**Comment by ADedecker** [comment]
This won't compile against master due to auto implicit.

bors cancel
bors d+

File: Mathlib/Data/List/EditDistance/Defs.lean [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+  · case cons y ys ih =>
+    dsimp [suffixLevenshtein]
+    rw [impl_length]
+    exact ih
+
+-- This is only used in keeping track of estimates.
+theorem suffixLevenshtein_eq (xs : List α) (y ys) {d} (w : d = suffixLevenshtein C xs ys) :
+    impl C xs y d = suffixLevenshtein C xs (y :: ys) := by
+  subst d
+  rfl
```
Comment (eric-wieser) [inline Mathlib/Data/List/EditDistance/Defs.lean:158]: What's going on here? Why  not state it as
```suggestion
-- This is only used in keeping track of estimates.
theorem suffixLevenshtein_eq (xs : List α) (y ys) :
    impl C xs y (suffixLevenshtein C xs ys) = suffixLevenshtein C xs (y :: ys) := by
  rfl
```

File: Mathlib/Data/List/EditDistance/Defs.lean [hunk 7]
```diff
@@ -0,0 +1,290 @@
+/-
+Copyright (c) 2023 Kim Liesinger. All rights reserved.
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/Data/List/EditDistance/Defs.lean:2]: If this is correct (and not a pseudonym/alias), then probably the PR description should gain a Co-authored-by line

**Comment by eric-wieser** [comment]
> This won't compile against master due to auto implicit.

I pushed the four character fix, but wonder if the offending lemma should exist at all (see my comment above).

**Comment by eric-wieser** [comment]
@semorrison, it looks like you ran a rebase with the wrong arguments; can you revert the last few commits and try again?

**Comment by semorrison** [comment]
It was an automatic rebase, no arguments from me! :-) In any case, I'd already merged this branch up into all the dependent PRs before seeing this, so I think it's done now. I'll happily share the credit!

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
bors seems to be confused on this one.

bors r-

**Pull request closed without merging**